### PR TITLE
Add midiproc from Crispy Doom

### LINKED ISSUE #90: audio: midiproc process for Windows
Win 7 x64 Pro SP1, Woof 1.2.2.

I'm sure you're well familiar with the symptoms. 
Crispy has crispy-midiproc.exe, what about Woof?

Great port otherwise, no need for Crispy Boom now, eh?

### PR DESCRIPTION
**Pull request by rfomin**
Fix #90

### CODE REVIEW

File: Source/i_sound.c [hunk 1]
```diff
@@ -34,8 +34,10 @@
```
Comment (fabiangreffrath) [inline Source/i_sound.c:34]: The changes to `i_sound.c` are massive. Please note that I took great care to only carefully modify this file and keep its original code structure intact as far as possible. I am doing this out of respect for the original authors of MBF and WinMBF, respectively. 

Whatever, it seems to me that you brute-forced the entire implementation from Crispy Doom into this file. I'd prefer if you could find out which changes are actually needed to get midiproc running and apply only those.

File: Source/i_sound.h [hunk 2]
```diff
@@ -97,7 +97,7 @@ void I_PauseSong(int handle);
 void I_ResumeSong(int handle);
 
 // Registers a song handle to song data.
-int I_RegisterSong(void *data, int size);
+int I_RegisterSong(void *data, int len);
```
Comment (fabiangreffrath) [inline Source/i_sound.h:100]: The name of the variable should be irrelevant here.

File: Source/mus2mid.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
+//
```
Comment (fabiangreffrath) [inline Source/mus2mid.h:1]: Is it really necessary to replace the entire `mmus2mid.c` by `mus2mid.c` or is there an isolated change that may be sufficient to apply?

File: Source/net_defs.h [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    int loadgame;
+    int random;  // [Strife only]
+
+    // These fields are only used by the server when sending a game
+    // start message:
+
+    int num_players;
+    int consoleplayer;
+
+    // Hexen player classes:
```
Comment (fabiangreffrath) [inline Source/net_defs.h:214]: Please clean these files off of everything that is not needed to get midiproc running.

File: midiproc/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+if(WIN32)
+	include(WoofSettings)
```
Comment (fabiangreffrath) [inline midiproc/CMakeLists.txt:2]: This has its indentation mixed up.

File: toolsrc/bmp2c.c [hunk 6]
```diff
@@ -6,6 +6,14 @@
 
 #define XPARENT (0xf7)	/* palette index representing transparent in BMP */
```
Comment (fabiangreffrath) [inline toolsrc/bmp2c.c:8]: This change is completely unrelated!

I appreciate your changes, but please file them as a separate PR. Also, there is already a setup for packed structs in https://github.com/fabiangreffrath/woof/blob/f0c5d0733f005b97cf7e69055573ef49914df7e4/Source/doomtype.h#L91, so please use this instead.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Hi @rfomin,

first, thank you very much for taking the time to implement this for Woof!

However, I've got quite some complaints about the way this has been integrated into the current code base. Please don't feel discouraged, though, your PR is highly appreciated.

Please take some more time to clean this up before we can proceed, thanks!

**Comment by rfomin** [comment]
I tried to restore most of the original code. To be honest I feel the temptation to remove the entire Woof sound system and replace it with Crispy/Chocolate one. Is it important to preserve such things?

File: Source/CMakeLists.txt [hunk 7]
```diff
@@ -42,7 +43,9 @@ set(WOOF_SOURCES
     m_misc2.c  m_misc2.h
     m_random.c m_random.h
     m_swap.h
-    mmus2mid.c mmus2mid.h
+    mmus2mid.c  mmus2mid.h
```
Comment (fabiangreffrath) [inline Source/CMakeLists.txt:46]: Here's a whitespace glitch.

File: Source/i_sound.c [hunk 8]
```diff
@@ -713,19 +738,38 @@ void I_PlaySong(int handle, int looping)
...
 //
+
+static int current_midi_volume;
+
 void I_SetMusicVolume(int volume)
 {
    // haleyjd 09/04/06: adjust to use scale from 0 to 15
-   Mix_VolumeMusic((volume * 128) / 15);
-}
+   int current_midi_volume = (volume * 128) / 15;
```
Comment (fabiangreffrath) [inline Source/i_sound.c:747]: Now this is declared in the scope of the function.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Much better now, thank you very much! Now I need to actually test the code on a Windows system...

**Comment by fabiangreffrath** [comment]
> To be honest I feel the temptation to remove the entire Woof sound system and replace it with Crispy/Chocolate one. Is it important to preserve such things?

It's probably not objectively important, but it's my coding philosophy. I prefer gradual improvements over "rip out and replace". Whenever I base my work on someone else's code, I try to preserve as much of the original code as possible. Because, there is always room for improvement, code is never hopelessly bad - and I don't wont to leave that impression over the original author's code.

When I started, the original WinMBF sound code would crash apparently randomly every few minutes. Now I've got it working stable, I have SFX pre-caching and pitch tuning - all by incremental improvements of the original code. And I am proud of this, it would be sad to rip all this effort out, even if this means more implementation work for contributors like you. I hope you understand!

Thank you again!

**Comment by rfomin** [comment]
Thanks for the explanation. I understand and appreciate your work, Crispy and Woof are my favorite ports. Hope I can help!

**Comment by fabiangreffrath** [comment]
> Thanks for the explanation. I understand and appreciate your work, Crispy and Woof are my favorite ports. 

That's nice to know, thank you!

> Hope I can help!

Well, I have just filed a bunch of wishlist items in the issue tracker, just saying... 😉

File: Source/i_midipipe.c [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+
+    if (!SetHandleInformation(midi_process_out_reader, HANDLE_FLAG_INHERIT, 0))
+    {
+        DEBUGOUT("Could not disinherit midiproc stdin");
+        return false;
+    }
+
+    // Define the command line.  Version, Sample Rate, and handles follow
+    // the executable name.
+    int snd_samplerate = 44100;
```
Comment (fabiangreffrath) [inline Source/i_midipipe.c:476]: No variable declaration in code, please.

File: midiproc/main.c [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+    {
+        MessageBox(NULL, TEXT("This program is tasked with playing Native ")
+                   TEXT("MIDI music, and is intended to be launched by ")
+                   TEXT(PROJECT_NAME) TEXT("."),
+                   TEXT(PROJECT_STRING), MB_OK | MB_ICONASTERISK);
+
+        return EXIT_FAILURE;
+    }
+
+    // Make sure our Choccolate Doom and midiproc version are lined up.
```
Comment (fabiangreffrath) [inline midiproc/main.c:409]: Why have you commented this out?

File: Source/i_midipipe.c (hunk 9 above)
Comment (fabiangreffrath) [reply Source/i_midipipe.c:476]: And why do we have to hard-code this here?

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]

**Comment by fabiangreffrath** [comment]
Also, `woof-midiproc.exe` is missing from the install and package rules.

File: Source/i_midipipe.c (hunk 9 above)
Comment (rfomin) [reply Source/i_midipipe.c:476]: snd_samplerate is a config parameter in Crispy Doom, should I replace it here with a constant?

Comment (fabiangreffrath) [reply Source/i_midipipe.c:476]: Ah, right, it's even hard-coded to that value in Woof's own `i_sound.c`.

Comment (fabiangreffrath) [reply Source/i_midipipe.c:476]: Anyway, please move the variable declaration so it doesn't appear in the middle of the code.

**Comment by rfomin** [comment]
> `woof-midiproc.exe` is missing from the install and package rules.
Does Crispy have these rules? I'm not good at CMake, not sure if install rules work at all on Windows.

File: midiproc/main.c (hunk 10 above)
Comment (rfomin) [reply midiproc/main.c:409]: Fix in https://github.com/fabiangreffrath/woof/pull/100/commits/33e2d0c750b38b95af6d37f420aa86a53363b7e4

**Comment by fabiangreffrath** [comment]
> Does Crispy have these rules? I'm not good at CMake, not sure if install rules work at all on Windows.

Me neither. 😉  I think you'll need to have a look at the last few lines in `Source/CMakeList.txt`.

**Comment by fabiangreffrath** [comment]
To provide some rationale why this must work, I create release tarballs with the `make package` rule, just as the `.travis.sh` script does.

**Comment by rfomin** [comment]
I couldn't figure it out quickly. Also, I will probably need MSYS or Linux for testing, so not today, sorry.

**Comment by fabiangreffrath** [comment]
No problem, we are not in a hurry!

**Comment by rfomin** [comment]
Tested it with MSYS. Also on Windows, you can make a package with `cpack -G ZIP` command.

**Review by fabiangreffrath — APPROVED** [review]
Very nice. So, do you consider this ready for merging now?

**Comment by rfomin** [comment]
Well, it works on my system. I think it's done.

**Comment by fabiangreffrath** [comment]
Well then, thank you very much!
